File: src/github/repos.rs
use std::env;

use anyhow::{Context, Result};
use chrono::{DateTime, Utc};
use reqwest::{header::USER_AGENT, Client};
use serde::{Deserialize, Serialize};

const GITHUB_TOKEN: &str = "GITHUB_ACCESS_TOKEN";

pub mod raw {
    use chrono::{DateTime, Utc};
    use serde::{Deserialize, Serialize};

    #[derive(Debug, PartialEq, Deserialize, Serialize, Clone)]
    pub struct Response {
        pub data: Data,
    }

    #[derive(Debug, PartialEq, Deserialize, Serialize, Clone)]
    pub struct Data {
        pub viewer: Viewer,
    }

    #[derive(Debug, PartialEq, Deserialize, Serialize, Clone)]
    pub struct Viewer {
        #[serde(rename = "pinnedItems")]
        pub pinned_items: PinnedItems,
    }

    #[derive(Debug, PartialEq, Deserialize, Serialize, Clone)]
    pub struct PinnedItems {
        pub nodes: Vec<Repository>,
    }

    #[derive(Debug, PartialEq, Deserialize, Serialize, Clone)]
    pub struct Repository {
        pub name: String,
        pub owner: Owner,
        #[serde(rename = "primaryLanguage")]
        pub primary_language: PrimaryLanguage,
        pub description: String,
        #[serde(rename = "updatedAt")]
        pub updated_at: DateTime<Utc>,
        #[serde(rename = "stargazerCount")]
        pub stargazer_count: u32,
    }

    #[derive(Debug, PartialEq, Deserialize, Serialize, Clone)]
    pub struct Owner {
        pub login: String,
    }

    #[derive(Debug, PartialEq, Deserialize, Serialize, Clone)]
    pub struct PrimaryLanguage {
        pub name: String,
        pub color: String,
    }
}

#[derive(Debug, PartialEq, Deserialize, Serialize, Clone)]
pub struct Repository {
    pub name: String,
    pub owner: String,
    pub language: String,
    pub language_color: String,
    pub description: String,
    pub updated_at: DateTime<Utc>,
    pub stargazers: u32,
}

pub async fn fetch_pinned_repos(client: &Client) -> Result<Vec<Repository>> {
    let resp = client
        .post("https://api.github.com/graphql")
        .bearer_auth(env::var(GITHUB_TOKEN).context("loading GITHUB access token failed")?)
        .header(USER_AGENT, "lcp/1.0")
        .body(r#"{"query": "query{viewer{pinnedItems(first:6,types:REPOSITORY){nodes{... on Repository{name owner{login}primaryLanguage{name color}description updatedAt stargazerCount isPrivate}}}}}","variables": {}}"#,
        )
        .send()
        .await
        .context("sending request to GraphQL API for recent games failed")?;
    let resp_text = resp
        .text()
        .await
        .context("getting raw response text failed")?;
    let mut raw_data: raw::Response = serde_json::from_str(&resp_text)
        .context(format!(
            "reading json failed from request to get data from graphql api: {}",
            resp_text
        ))
        .context("reading json failed from request to get pinned repos")?;
    Ok(raw_data
        .data
        .viewer
        .pinned_items
        .nodes
        .iter_mut()
        .map(|a| Repository {
            name: a.name.to_owned(),
            owner: a.owner.login.to_owned(),
            language: a.primary_language.name.to_owned(),
            language_color: a.primary_language.color.to_owned(),
            description: a.description.to_owned(),
            updated_at: a.updated_at,
            stargazers: a.stargazer_count,
        })
        .collect())
}

File: internal/apis/github/repos.go
package github

import (
	"context"
	"errors"
	"fmt"
	"net"
	"strings"
	"syscall"

	"github.com/shurcooL/githubv4"
	"go.mattglei.ch/lcp/internal/apis"
	"go.mattglei.ch/lcp/pkg/lcp"
	"go.mattglei.ch/timber"
)

type pinnedItemsQuery struct {
	Viewer struct {
		PinnedItems struct {
			Nodes []struct {
				Repository struct {
					Name  githubv4.String
					Owner struct {
						Login githubv4.String
					}
					PrimaryLanguage struct {
						Name  githubv4.String
						Color githubv4.String
					}
					Description githubv4.String
					UpdatedAt   githubv4.DateTime
					IsPrivate   githubv4.Boolean
					ID          githubv4.ID
					URL         githubv4.URI
				} `graphql:"... on Repository"`
			}
		} `graphql:"pinnedItems(first: 6, types: REPOSITORY)"`
	}
}

func fetchPinnedRepos(client *githubv4.Client) ([]lcp.GitHubRepository, error) {
	var query pinnedItemsQuery
	err := client.Query(context.Background(), &query, nil)
	if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
		timber.Warning(cacheInstance.LogPrefix(), "connection timed out for getting pinned repos")
		return []lcp.GitHubRepository{}, apis.ErrWarning
	}
	if err != nil && (errors.Is(err, syscall.ECONNRESET) ||
		strings.Contains(err.Error(), "connection reset by peer")) {
		timber.Warning(cacheInstance.LogPrefix(),
			"connection reset by peer while getting pinned repos")
		return []lcp.GitHubRepository{}, apis.ErrWarning
	}
	if err != nil {
		return nil, fmt.Errorf("%w querying github's graphql API failed", err)
	}

	var repositories []lcp.GitHubRepository
	for _, node := range query.Viewer.PinnedItems.Nodes {
		repositories = append(repositories, lcp.GitHubRepository{
			Name:          string(node.Repository.Name),
			Owner:         string(node.Repository.Owner.Login),
			Language:      string(node.Repository.PrimaryLanguage.Name),
			LanguageColor: string(node.Repository.PrimaryLanguage.Color),
			Description:   string(node.Repository.Description),
			UpdatedAt:     node.Repository.UpdatedAt.Time,
			ID:            fmt.Sprint(node.Repository.ID),
			URL:           fmt.Sprint(node.Repository.URL.URL),
		})
	}
	return repositories, nil
}
